File: samples/guides2/src/main/java/app/cash/tempest2/guides/java/AsynchronousProgramming.java
package app.cash.tempest2.guides.java;

import app.cash.tempest2.musiclibrary.java.AsyncMusicTable;
import app.cash.tempest2.musiclibrary.java.PlaylistInfo;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import software.amazon.awssdk.enhanced.dynamodb.Expression;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;

public class AsynchronousProgramming {

  private final AsyncMusicTable table;

  public AsynchronousProgramming(AsyncMusicTable table) {
    this.table = table;
  }

  public CompletableFuture<Void> changePlaylistName(String playlistToken, String newName) {
    // Read.
    return table.playlistInfo()
        .loadAsync(new PlaylistInfo.Key(playlistToken))
        .thenCompose(existing -> {
          if (existing == null) {
            throw new IllegalStateException("Playlist does not exist: " + playlistToken);
          }
          // Modify.
          PlaylistInfo newPlaylist = new PlaylistInfo(
              existing.playlist_token,
              newName,
              existing.playlist_tracks,
              // playlist_version.
              existing.playlist_version + 1
          );
          // Write.
          return table.playlistInfo()
              .saveAsync(
                  newPlaylist,
                  ifPlaylistVersionIs(existing.playlist_version)
              );
        });
  }

  private Expression ifPlaylistVersionIs(Long playlist_version) {
    return Expression.builder()
        .expression("playlist_version = :playlist_version")
        .expressionValues(
            Map.of(":playlist_version", AttributeValue.builder().n("" + playlist_version).build()))
        .build();
  }
}

File: samples/guides2/src/main/kotlin/app/cash/tempest2/guides/AsynchronousProgramming.kt
package app.cash.tempest2.guides

import app.cash.tempest2.musiclibrary.AsyncMusicTable
import app.cash.tempest2.musiclibrary.PlaylistInfo
import software.amazon.awssdk.enhanced.dynamodb.Expression
import software.amazon.awssdk.services.dynamodb.model.AttributeValue

class AsynchronousProgramming(
  private val table: AsyncMusicTable,
) {

  suspend fun changePlaylistName(playlistToken: String, newName: String) {
    // Read.
    val existing = checkNotNull(
      table.playlistInfo.load(PlaylistInfo.Key(playlistToken)) // This is a suspend function.
    ) { "Playlist does not exist: $playlistToken" }
    // Modify.
    val newPlaylist = existing.copy(
      playlist_name = newName,
      playlist_version = existing.playlist_version + 1
    )
    // Write.
    table.playlistInfo.save( // This is a suspend function.
      newPlaylist,
      ifPlaylistVersionIs(existing.playlist_version)
    )
  }

  private fun ifPlaylistVersionIs(playlist_version: Long): Expression {
    return Expression.builder()
      .expression("playlist_version = :playlist_version")
      .expressionValues(mapOf(":playlist_version" to AttributeValue.builder().n("$playlist_version").build()))
      .build()
  }
}